service/src/java/org/apache/hive/service/cli/thrift/ThriftHttpServlet.java

```diff
@@ -284,7 +283,7 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
       }
       // Send a 401 to the client
       response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
-      if(isAuthTypeEnabled(request, HiveAuthConstants.AuthTypes.KERBEROS)) {
+      if(authType.isEnabled(HiveAuthConstants.AuthTypes.KERBEROS)) {
```

Comment:
Is there any reason to change this check to `authType.isEnabled(Kerberos)`?
I wondering if the HS2 supports both Kerberos and LDAP, if the current request for LDAP authentication is failed, we don't need to send `HttpAuthUtils.WWW_AUTHENTICATE` header back.

Reply:
You are right. I wanted to fix the case where no `Authorization` header is received and Kerberos authentication is enabled. In this case, `isAuthTypeEnabled(request, HiveAuthConstants.AuthTypes.KERBEROS)` will always be false, i.e., a `WWW-Authenticate` header will never be returned in an [rfc4559 4.1](https://datatracker.ietf.org/doc/html/rfc4559#section-4.1) compliant manner. (Note: In the current code this works fine, because if the first allowed auth mode is Kerberos, then the above `isAuthTypeEnabled(...)` condition will be true.)
Maybe I should add an additional check to the condition that there is no `Authorization` header in the request. What do you think about this?

Reply:
The HttpKerberosRequestInterceptor in the client will add the `AUTHORIZATION` header for every http request:
https://github.com/apache/hive/blob/master/jdbc/src/java/org/apache/hive/jdbc/HttpKerberosRequestInterceptor.java#L63-L65
so I think for a valid HiveConnection request, there should be a `AUTHORIZATION` header in the request.

Reply:
@dengzhhu653 
It seems that this change is necessary, because the Knox gateway first calls hs2 without _Authorization_ header, then if there is a _WWW-Authenticate: Negotiate_ header in the response, it will only initiate a normal authentication with _Authorization_ header ([so it implements reactive authentication](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-authenticator/#reactive-authentication)):
https://github.com/apache/knox/blob/master/gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/auth/SpnegoAuthInterceptor.java#L105-L113 
The current code now doesn't work well behind Knox gw when the request is not coming directly from the jdbc driver, so I'll create a bug ticket for this problem and submit the fix.